File: src/main/java/com/kamikazejam/datastore/base/field/FieldWrapper.java
package com.kamikazejam.datastore.base.field;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.kamikazejam.datastore.base.Store;
import lombok.Getter;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Map;
import java.util.Objects;

/**
 * A wrapper for fields that enforces access control based on Store state
 */
@SuppressWarnings("unchecked")
public class FieldWrapper<T> {
    private @Nullable T value;
    @Getter
    private final @Nullable T defaultValue;
    @JsonIgnore
    private @Nullable Store<?, ?> parent;
    private @NotNull final String name;
    private @NotNull final Class<?> valueType;

    private FieldWrapper(@NotNull String name, @Nullable T defaultValue, @NotNull Class<T> valueType) {
        this(name, defaultValue, valueType, null);
    }

    @SuppressWarnings("unused")
    private FieldWrapper(@NotNull String name, @Nullable T defaultValue, @NotNull Class<?> valueType, @Nullable Class<?> elementType) {
        this.name = name;
        this.value = defaultValue;
        this.defaultValue = defaultValue;
        this.valueType = valueType;
        // We don't actually need `elementType` in our implementation yet, but we
        //  keep it in case it is needed in the future
    }


    // ------------------------------------------------------ //
    // Static Constructors                                    //
    // ------------------------------------------------------ //

    public static <T> FieldWrapper<T> of(@NotNull String name, @Nullable T defaultValue, @NotNull Class<T> valueType) {
        return new FieldWrapper<>(name, defaultValue, valueType);
    }

    // Generic constructor for any collection type
    public static <C extends Collection<E>, E> FieldWrapper<C> ofColl(
            @NotNull String name,
            @Nullable C defaultValue,
            @NotNull Class<? super C> collectionType
    ) {
        return new FieldWrapper<>(name, defaultValue, collectionType, null);
    }

    // Generic constructor for any map type
    public static <K, V, M extends Map<K,V>> FieldWrapper<M> ofMap(
            @NotNull String name,
            @Nullable M defaultValue,
            @NotNull Class<? super M> mapType
    ) {
        return new FieldWrapper<>(name, defaultValue, mapType, null);
    }

    @NotNull
    public String getName() {
        return name;
    }

    @NotNull
    public Class<T> getValueType() {
        return (Class<T>) valueType;
    }

    @ApiStatus.Internal
    public void setParent(@NotNull Store<?, ?> parent) {
        this.parent = parent;
    }

    public T get() {
        if (parent == null) { throw new IllegalStateException("[FieldWrapper#get] Field not registered with a parent document"); }
        return value;
    }

    public void set(T value) {
        if (parent == null) { throw new IllegalStateException("[FieldWrapper#set] Field not registered with a parent document"); }
        if (parent.isReadOnly()) { throw new IllegalStateException("Cannot modify field '" + name + "' in read-only mode");}
        this.value = value;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof FieldWrapper<?> that)) return false;
        return Objects.equals(value, that.value) && Objects.equals(name, that.name) && Objects.equals(valueType, that.valueType);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, name, valueType);
    }

    @Override
    public String toString() {
        return "FieldWrapper{name='" + name + "', value=" + value + ", valueType=" + valueType + "}";
    }
}

File: src/main/kotlin/com/kamikazejam/datastore/base/field/FieldWrapper.kt
@file:Suppress("unused")

package com.kamikazejam.datastore.base.field

import com.google.common.base.Preconditions
import com.kamikazejam.datastore.base.Store
import com.kamikazejam.datastore.base.data.StoreData
import org.jetbrains.annotations.ApiStatus
import java.util.*

sealed interface FieldWrapper<D : StoreData<Any>> : FieldProvider {
    val name: String
    val isWriteable: Boolean
    // we need a creator function that can create an instance of D
    val creator: () -> D
    @set:ApiStatus.Internal
    var parent: Store<*, *>?

    fun getData(): D?

    fun setDataNotNull(data: D)

    override val fieldWrapper: FieldWrapper<*>
        get() = this

    override fun equals(other: Any?): Boolean
    override fun hashCode(): Int
}

sealed interface RequiredField<D : StoreData<Any>> : FieldWrapper<D> {
    val defaultValue: D
    fun setData(data: D)
    override fun setDataNotNull(data: D) = setData(data)

    override fun getData(): D

    companion object {
        @JvmStatic
        fun <D : StoreData<Any>> of(name: String, defaultValue: D, creator: () -> D = { defaultValue }): RequiredField<D> =
            RequiredFieldImpl(name, defaultValue, creator)
    }
}

sealed interface OptionalField<D : StoreData<Any>> : FieldWrapper<D> {
    val defaultValue: D?
    fun getDataOrDefault(default: D): D
    fun setData(data: D?)
    override fun setDataNotNull(data: D) = setData(data)

    override fun getData(): D?

    companion object {
        @JvmStatic
        fun <D : StoreData<Any>> of(name: String, defaultValue: D?, creator: () -> D): OptionalField<D> =
            OptionalFieldImpl(name, defaultValue, creator)
    }
}

@Suppress("DuplicatedCode")
private class RequiredFieldImpl<D : StoreData<Any>>(
    override val name: String,
    override val defaultValue: D,
    override val creator: () -> D = { defaultValue }
) : RequiredField<D> {
    private var data: D = defaultValue
    override var parent: Store<*, *>? = null
        set(value) {
            field = value
            data.setParent(value)
        }

    override fun getData(): D {
        Preconditions.checkState(parent != null, "[RequiredField#get] Field not registered with a parent document")
        Preconditions.checkState(data.parent != null, "[RequiredField#get] Data not registered with a parent document")
        return data
    }

    override fun setData(data: D) {
        Preconditions.checkState(isWriteable, "Cannot modify field '$name' in read-only mode")
        Preconditions.checkState(
            defaultValue::class.isInstance(data),
            "Data $data (${data.let { it::class.java }}) is not an instance of the field type (${defaultValue::class.java})"
        )
        this.data = data
        this.data.setParent(parent)
    }

    override val isWriteable: Boolean
        get() {
            val p = this.parent
            Preconditions.checkState(
                p != null,
                "[RequiredField#isWriteable] Field not registered with a parent document"
            )
            checkNotNull(p)
            return !p.readOnly && data.isWriteable
        }

    override fun equals(other: Any?): Boolean {
        if (other !is RequiredFieldImpl<*>) return false
        return data == other.data && name == other.name
    }

    override fun hashCode(): Int {
        return Objects.hash(data, name)
    }
}

private class OptionalFieldImpl<D : StoreData<Any>>(
    override val name: String,
    override val defaultValue: D?,
    override val creator: () -> D
) : OptionalField<D> {
    private var data: D? = defaultValue
    override var parent: Store<*, *>? = null
        set(value) {
            field = value
            data?.setParent(value)
        }

    override fun getData(): D? {
        val d = this.data
        Preconditions.checkState(parent != null, "[OptionalField#get] Field not registered with a parent document")
        Preconditions.checkState((d == null || d.parent != null), "[OptionalField#get] Data not registered with a parent document")
        return d
    }

    override fun getDataOrDefault(default: D): D {
        return getData() ?: default
    }

    override fun setData(data: D?) {
        Preconditions.checkState(isWriteable, "Cannot modify field '$name' in read-only mode")
        this.data = data
        this.data?.setParent(parent)
    }

    override val isWriteable: Boolean
        get() {
            val p = this.parent
            Preconditions.checkState(
                p != null,
                "[OptionalField#isWriteable] Field '$name' not registered with a parent document"
            )
            checkNotNull(p)
            return !p.readOnly && (data?.isWriteable ?: true)
        }

    override fun equals(other: Any?): Boolean {
        if (other !is OptionalFieldImpl<*>) return false
        return data == other.data && name == other.name
    }

    override fun hashCode(): Int {
        return Objects.hash(data, name)
    }
}